### ISSUE TITLE
Issue in saving Many to One relation (UNIQUE constraint failed)

### ISSUE THREAD

**Issue by hamzaazad** [typeorm/typeorm#1794]
I have the following entities:

User:

    @Entity('user')
    export class User extends BaseEntity {
     
    @Column()
    username: string;

    @Column()
    firstName: string;

    @Column({ nullable: true })
    lastName: string;

    @Column({ nullable: true })
    description: string;

    @Column()
    email: string;

    @Column({ nullable: true })
    profileImage: number;

    @OneToMany(type => Ranch, ranch => ranch.owner)
    ranches: Ranch[];

    }

Ranch:

    @Entity('ranch')
    export class Ranch extends BaseEntity {
    
    @Column()
    name: string;

    @Column({ nullable: true })
    address: string;

    @Column({ nullable: true })
    area: number;

    @Column({ nullable: true })
    contact: string;

    @Column({ nullable: true })
    description: string;

    @Column({ nullable: true })
    state: number;

    @Column()
    type: number;

    @Column({ nullable: true, default: 0 })
    cattleCount: number;

    @Column({ nullable: true, default: 0 })
    workerCount: number;

    @ManyToOne(type => User, owner => owner.ranches, {
        cascadeInsert: true,
        cascadeUpdate: true,
        eager: true
    })
    owner: User;

    }

Base Entity:

    @Entity('base_entity')
    export class BaseEntity {
    
    @PrimaryColumn()
    id: number;

    @Column()
    dateCreated: number;

    @Column({ nullable: true })
    dateModified: number;
    
    }

This is the json that I am using to save ranch list using one save call like mentioned in the TypOrm ManyToOne OneToMany page:

    {
        "data": {
        "content": [
            {
            "id": 1,
            "dateCreated": 1520967600000,
            "dateModified": null,
            "name": "AAA System Ranch",
            "description": null,
            "address": "AAA Head Office",
            "contact": null,
            "type": 1,
            "area": null,
            "user": {
                "dateCreated": 1520967600000,
                "dateModified": null,
                "id": 1,
                "username": "mmudassir@folio3.com",
                "firstName": "admin",
                "lastName": null,
                "email": "mmudassir@folio3.com"
            },
            "cattleCount": null,
            "workerCount": 0
            },
            {
            "id": 7,
            "dateCreated": 1521630736000,
            "dateModified": 1521630736000,
            "name": "PSL Ranch",
            "description": null,
            "address": "shah latif town",
            "contact": "123456",
            "type": 1,
            "area": null,
            "user": {
                "dateCreated": 1520967600000,
                "dateModified": null,
                "id": 1,
                "username": "mmudassir@folio3.com",
                "firstName": "admin",
                "lastName": null,
                "email": "mmudassir@folio3.com"
            },
            "cattleCount": null,
            "workerCount": 0
            }
        ]
        }
    }

In order to save the ranch list I do the following:

           const ranchData = { _RANCHES_JSON_ }
           const ranchObjects = this.ranchRepository.create(ranchData);
            this.ranchRepository.save(ranchObjects).then((data) => {
                // success
            }).catch((err) => {
                // failure
            });

The first time this code runs, I receive the follwing error:

    SQLError {code: 6, message: "could not execute statement due to a constraint failure (19 UNIQUE         
    constraint failed: user.id)"}
    code
    :
    6
    message
    :
    "could not execute statement due to a constraint failure (19 UNIQUE constraint failed: user.id)"
    __proto__
    :
    SQLError

The next time the same code runs, no error is received. User is saved both times but since I receive this error, I do not display the list and display an error instead. What should I do?

I am using websql for testing and will use sqlite for ionic on device.

**Comment by pleerock** [comment]
UNIQUE constraint failed probably means you are attaching same object twice. Try to remove all cascades and start from clean and find your issue.

**Comment by pleerock** [comment]
Closing as no answer from reporter.

**Comment by akashbiz** [comment]
@pleerock , I am facing same issue. I tried removing cascade and inserting data it only inserts data(ranch)  without dependent data(user in above case). I want dependent to be inserted first which has primaryColumn as id which I am receiving from our server. 
If I keep cascading insert ON, it should insert dependent data only once if not exist and if exist just update it. Right? But It seems it is trying insert data each time. Hence I am receiving 
**could not execute statement due to a constraint failure (19 UNIQUE constraint failed:**

**Comment by havenchyk** [comment]
@akashbiz feel free to open a new issue with more details and a small repo to reproduce the problem

**Comment by akashbiz** [comment]
Hey @havenchyk , Thanks for quick reply. Can you share any example of One to Many/Many to One relationship where cascade insert is used. It will be a great help for me. If it still didn't work out then I will surely create a new issue with some code which I have tried with.
Thanks,
Akash.

**Comment by havenchyk** [comment]
@akashbiz have you checked https://github.com/typeorm/typeorm/blob/master/docs/many-to-one-one-to-many-relations.md?

**Comment by havenchyk** [comment]
@akashbiz then create a new issue to see your used code
